Manual/Iterators.lean

```diff
@@ -0,0 +1,1123 @@
+def allNats : List Nat :=
+  let steps : Iter Nat := (0...*).iter
+  steps.allowNontermination.toList
+```
+
+:::
+
+:::example "Consuming Iterators in Loops"
+```imports -show
+import Std.Data.Iterators
+```
+```lean -show
+open Std.Iterators
+```
+This program creates an iterator of strings from a range, and then consumes the strings in a {keywordOf Lean.Parser.Term.doFor}`for` loop:
+```lean (name := iterFor)
+def countdown (n : Nat) : IO Unit := do
+  let steps : Iter String := (0...n).iter.map (s!"{n - ·}!")
+  for i in steps do
+    IO.println i
+  IO.println "Blastoff!"
+
+#eval countdown 5
+```
+```leanOutput iterFor
+5!
+4!
+3!
+2!
+1!
+Blastoff!
+```
+:::
+
+:::example "Consuming Iterators Directly"
+```imports -show
+import Std.Data.Iterators
+```
+```lean -show
+open Std.Iterators
+```
+The function {name}`countdown` calls the range iterator's {name Iter.step}`step` function directly, handling each of the three possible cases.
+```lean
+def countdown (n : Nat) : IO Unit := do
+  let steps : Iter Nat := (0...n).iter
+  go steps
+where
+  go iter := do
+    match iter.step with
+    | .done _ => pure ()
+    | .skip iter' _ => go iter'
+    | .yield iter' i _ => do
+      IO.println s!"{i}!"
+      if i == 2 then
+        IO.println s!"Almost there..."
+      go iter'
+  termination_by iter.finitelyManySteps
+```
+:::
+
+## Stepping Iterators
+
+Iterators are manually stepped using {name}`Iter.step` or {name}`IterM.step`.
+
+{docstring Std.Iterators.Iter.step}
+
+{docstring Std.Iterators.IterM.step}
+
+### Termination
+
+When manually stepping an finite iterator, the termination measures {name Iter.finitelyManySteps}`finitelyManySteps` and {name Iter.finitelyManySkips}`finitelyManySkips` can be used to express that each step brings iteration closer to the end.
+The proof automation for {ref "well-founded-recursion"}[well-founded recursion] is pre-configured to prove that recursive calls after steps reduce these measures.
+
+:::example "Finitely Many Skips"
+```imports -show
+import Std.Data.Iterators
+```
+```lean -show
+open Std.Iterators
+```
+This function returns the first element of an iterator, if there is one, or {name}`none` otherwise.
+Because the iterator must be productive, it is guaranteed to return an element after at most a finite number of {name PlausibleIterStep.skip}`skip`s.
+This function terminates even for infinite iterators.
+```lean
+def getFirst [Iterator α Id β] [Productive α Id]
+    (it : @Iter α β) : Option β :=
+  match it.step with
+  | .done .. => none
+  | .skip it' .. => getFirst it'
+  | .yield _ x .. => pure x
+termination_by it.finitelyManySkips
+```
+:::
+
+{docstring Iter.finitelyManySteps}
+
+{docstring IterM.finitelyManySteps}
+
+{docstring IterM.TerminationMeasures.Finite +allowMissing}
+
+{docstring Iter.finitelyManySkips}
+
+{docstring IterM.finitelyManySkips}
+
+{docstring IterM.TerminationMeasures.Productive +allowMissing}
+
+{docstring Iter.attachWith}
+
+{docstring IterM.attachWith}
```

Comment:
`attachWith` doesn't really help with termination proofs. Perhaps these are better placed among the combinators?

Reply:
Ah, I'd thought of the way that `attach` is useful to help wf-rec go through in nested inductive types. I'll go move them.